akka-http-core/src/main/scala/akka/http/impl/engine/http2/framing/FrameRenderer.scala

```diff
@@ -24,141 +21,200 @@ private[http2] object FrameRenderer {
   def render(frame: FrameEvent): ByteString =
     frame match {
       case GoAwayFrame(lastStreamId, errorCode, debug) =>
-        val bb = new ByteStringBuilder
-        bb.putInt(lastStreamId)
-        bb.putInt(errorCode.id)
-        // appends debug data, if any
-        bb.append(debug)
-
-        renderFrame(
-          Http2Protocol.FrameType.GOAWAY,
-          Http2Protocol.Flags.NO_FLAGS,
-          Http2Protocol.NoStreamId,
-          bb.result()
-        )
+        HeaderBuilder(8 + debug.size)
+          .putHeader(
+            Http2Protocol.FrameType.GOAWAY,
+            Http2Protocol.Flags.NO_FLAGS,
+            Http2Protocol.NoStreamId,
+          )
+          .putInt32(lastStreamId)
+          .putInt32(errorCode.id)
+          // appends debug data, if any
+          .put(debug)
+          .result()
 
       case DataFrame(streamId, endStream, payload) =>
         // TODO: should padding be emitted? In which cases?
-
-        renderFrame(
-          Http2Protocol.FrameType.DATA,
-          Http2Protocol.Flags.END_STREAM.ifSet(endStream),
-          streamId,
-          payload
-        )
+        HeaderBuilder(payload.size)
+          .putHeader(
+            Http2Protocol.FrameType.DATA,
+            Http2Protocol.Flags.END_STREAM.ifSet(endStream),
+            streamId
+          )
+          .put(payload)
+          .result()
       case HeadersFrame(streamId, endStream, endHeaders, headerBlockFragment, prioInfo) =>
-        val renderedPrioInfo = prioInfo.map(renderPriorityInfo).getOrElse(ByteString.empty)
-
-        renderFrame(
-          Http2Protocol.FrameType.HEADERS,
-          Http2Protocol.Flags.END_STREAM.ifSet(endStream) |
-            Http2Protocol.Flags.END_HEADERS.ifSet(endHeaders) |
-            Http2Protocol.Flags.PRIORITY.ifSet(prioInfo.isDefined),
-          streamId,
-          renderedPrioInfo ++ headerBlockFragment
-        )
+        HeaderBuilder(headerBlockFragment.size + (if (prioInfo.isDefined) 5 else 0))
```

Comment:
Perhaps follow the order of the fields

```suggestion
        HeaderBuilder((if (prioInfo.isDefined) 5 else 0) + headerBlockFragment.size)
```

Reply:
:+1: done